bazaarvoice/emodb
# Add support for role namespaces

## Github Issue #

[93](https://github.com/bazaarvoice/emodb/issues/93)

## What Are We Doing Here?

As described in the corresponding issue, Emo is moving towards a system of delegated API key management.  To fully get there is a four step process:

1. Create the ability to restrict an API key from creating a role with permissions wider than the API key's current permissions (https://github.com/bazaarvoice/emodb/issues/63)
1. Add the ability to group roles together by namespace (https://github.com/bazaarvoice/emodb/issues/93)
1. Refactor role and API key management permissions to restrict role management and assignment by group
1. Expose role and API key management as a public REST interface (https://github.com/bazaarvoice/emodb/issues/50)

This PR addresses the second step in this process.  The major changes in this PR are:

1. Created first-order `Role` object and the associated management interface and implementations.  Each role belongs to exactly one group, including the implicit `null` group.
1. Split `AuthIdentityManager` and `PermissionManager` interfaces to inherit from more restrictive read-only `AuthIdentityReader` and `PermissionReader` interfaces, respectively.  This allows authenticators such as `ApiKeyRealm` to use the reader interfaces without risk of exposing the ability to update users, roles, or permissions.  This is not strictly necessary for this PR but with the added interfaces for manipulating security artifacts introduced in this PR it was a good time to close this potential gap.
1. Refactored `PermissionManager` to be more agnostic about the permissions it manages.  This was done to provide a clean separation of role and permission management since roles are now first-order objects.

Although this PR updates many files it's changes 2 and 3 above which account for a good share of those, due to the class and method name changes percolating through the system.

**Design Decisions**

It's possible that roles could have been modeled as belonging to multiple groups, not one exclusively.  However, this was not done for the following reasons:

- There is no use case for it.
- It complicates the goal of grouping roles for the purpose of managing permissions to use those roles.  If a role belongs to multiple groups the permissions associated with that role become much less clear.  For example, if a role belongs to "group1" and "group2" and a user has update permission for roles in "group1" but not "group2" then should he be able to update the role?
- If a role can belong to more than one group then the role needs a unique identifier which is separate from the role ID or group, such as a random UUID.  While this itself is not a bad thing it requires a complicated dereferencing system and upgrade path with no recognizable benefit.

The unique identifier for a role is derived from the group and a user-provided ID.  This means that if a role is deleted and then re-created with the same group and ID then any API keys who had been assigned the role prior to its deletion will still be assigned the role after it is re-created, assuming the role was not explicitly removed from the API key at any time during this process.  This is an intentional decision for the following reasons:

- It avoids potential confusion given that the (group, ID) is effectively a natural key since the role's creator provides both.
- It allows for the assignment of roles to be completely separated from the management of roles.  For example, deleting a role does not require scanning every API key and removing the role from that key if present.
- It simplifies role management by requiring neither maintaining multiple states for a role nor creating non-natural unique identifiers for new roles upon creation.  Both of these solutions add significant complexity for a small benefit which is arguably not the desired behavior.

Note that in addition to (group, ID) the role creator can also set a "name" and "description" attribute.  Both of these are informational only and there is no uniqueness guarantee on either of these fields, although it likely benefits the creator to assign unique names to roles within her group.

**Upgrade path**

The naming convention for roles is as follows:

- If the role has no group (group == `null`) then the role's persistence ID is just the role ID.
- If the role has a group then the role's persistence ID is "group/ID".

For this reason, prior to upgrading there must be no roles whose name includes the "/" character.  Any such roles should be migrated to new roles with no "/" prior to upgrading.

Once this is done all existing roles are implicitly grandfathered into the `null` group of roles.  All existing API keys and permissions are already persisted in a forward-compatible format.  Upgraded Emo servers can be deployed along with the existing ones with no down time or interruption of normal service.  The only restriction is that roles should not be modified while the upgrade is taking place.

While the current users, roles, and permissions are forward-compatible the ability to manage existing roles is not.  The `RebuildMissingRolesTask` must be run once per environment after the upgrade to populate the new role persistence tables.  Until this is done the administrator will not be able to modify any existing roles.

## How to Test and Verify

1. Check out this PR
2. Use the `RoleAdminTask` and `ApiKeyAdminTask` to create and assign multiple roles, both with and without a group.  Verify all works as expected.

## Risk

The greatest risk is an error which causes roles to become unstable, since this could lead to users being unable to access parts of Emo for which they should have permission.

### Level 

`Medium` While this seems like a large scale PR most of the changes are centered around the management of roles.  Permission management is intentionally unaffected to minimize the risk of any changes impacting the existing permissions system.

### Required Testing

`Manual`

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
